### PR TITLE
Ensure sample size limited for large data

### PR DESCRIPTION
**Pull request by JGSweets**
Address the issue: https://github.com/capitalone/DataProfiler/issues/171

Speeds are now below the original speed and the speed from the PR previous to PR #159 .

Profile time for 26MB file:
Current: 24+s
Prior to PR #159: ~8s
This PR: ~6.5s

### CODE REVIEW

File: dataprofiler/profilers/profile_builder.py [hunk 1]
```diff
@@ -421,27 +418,16 @@ def __init__(self, data, samples_per_update=None, min_true_samples=0,
...
-                    labeler_type='structured',
-                    dirpath=data_labeler_options.data_labeler_dirpath,
-                    load_options=None)
-                self.options.set({'data_labeler.data_labeler_object': data_labeler})
-                
-            except Exception as e:
-                utils.warn_on_profile('data_labeler', e)
-                self.options.set({'data_labeler.is_enabled': False})
-
-        self.update_profile(data)
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:444]: moved to inside `update_profile`, was done initially in case merging of profiles could duplicate the labeler, but looks like merge handles it by passing its options through.

Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:444]: Technically this allows a profile to be instantiated without creating the labeler until profile time making it faster. Not sure if this is desirable.

File: dataprofiler/profilers/profile_builder.py [hunk 2]
```diff
@@ -602,37 +604,32 @@ def update_profile(self, data, sample_size=None, min_true_samples=None):
...
+            self._update_profile_from_chunk(
+                data.data, sample_size, min_true_samples, self.options)
             self._update_row_statistics(data.data)
             self.encoding = data.file_encoding
             self.file_type = data.data_type
         elif isinstance(data, pd.DataFrame):
-            self._profile = self._update_profile_from_chunk(
-                data, self._profile, sample_size, min_true_samples, self.options)
+            self._update_profile_from_chunk(
+                data, sample_size, min_true_samples, self.options)
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:617]: previously `_update_profile_from_chunk` was static, but it doesn't really make sense and means we no longer have to pass profiles and can just use `self._profile` within the function

File: dataprofiler/profilers/profile_builder.py [hunk 3]
```diff
@@ -602,37 +604,32 @@ def update_profile(self, data, sample_size=None, min_true_samples=None):
...
         else:
             raise ValueError(
                 "Data must either be imported using the data_readers or "
                 "pd.DataFrame."
             )
-   
-    @staticmethod
-    def _update_profile_from_chunk(df, profile=None, sample_size=None,
+
+    def _update_profile_from_chunk(self, df, sample_size=None,
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:626]: previously `_update_profile_from_chunk` was static, but it doesn't really make sense and means we no longer have to pass profiles and can just use `self._profile` within the function

This allows us to get the sample_size within `_update_profile_from_chunk`

File: dataprofiler/profilers/profile_builder.py (hunk 2 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:617]: This was done so that `_get_sample_size` could be called within `_update_profile_from_chunk`. If we did it here, the function would have to be rearranged in order to account for data checks which previously were not needed.

Currently the data check is the `else` in the conditionals, but if sample size was acquired here, it would need to move to the top  followed by the `_get_sample_size` and the instance check conditionals. Feasible if desired.

File: dataprofiler/profilers/profile_builder.py [hunk 4]
```diff
@@ -642,14 +639,29 @@ def _update_profile_from_chunk(df, profile=None, sample_size=None,
         :return: list of column profile base subclasses
         :rtype: list(BaseColumnProfiler)
         """
-        
-        if not profile:
-            profile = OrderedDict()
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:647]: just use the `self._profile`

File: dataprofiler/profilers/profile_builder.py (hunk 3 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:626]: I'm no recalling the original intent was because if something error'ed within update, we wouldn't update anything. We can revert to keep this intent if desired.

File: dataprofiler/tests/profilers/profiler_options/test_profiler_options.py [hunk 5]
```diff
@@ -351,6 +350,7 @@ def test_data_labeler(self, *mocks):
         options.structured_options.data_labeler.data_labeler_dirpath \
             = "Test_Dirpath"
         options.structured_options.data_labeler.max_sample_size = 50
+        options.structured_options.multiprocess.is_enabled = False
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/profiler_options/test_profiler_options.py:353]: speeds up test

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 6]
```diff
@@ -288,13 +286,16 @@ def test_text_data_raises_error(self):
...
+    @mock.patch('dataprofiler.profilers.profile_builder.DataLabeler')
+    @mock.patch('dataprofiler.profilers.profile_builder.Profiler.'
+                '_update_row_statistics')
+    @mock.patch('dataprofiler.profilers.profile_builder.StructuredDataProfile')
     def test_sample_size_warning_in_the_profiler(self, *mocks):
+        # structure data profile mock
+        sdp_mock = mock.Mock()
+        sdp_mock.clean_data_and_get_base_stats.return_value = (dict(a=1), None)
+        mocks[0].return_value = sdp_mock
+
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:298]: speeds up test

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 7]
```diff
@@ -493,12 +493,31 @@ def test_null_count(self):
...
-
     def test_generating_report_ensure_no_error(self):
         file_path = os.path.join(test_root_path, 'data', 'csv/diamonds.csv')
         data = pd.read_csv(file_path)
         profile = dp.Profiler(data[:1000])
-        readable_report = profile.report(report_options={"output_format":"compact"})
+        readable_report = profile.report(
+            report_options={"output_format": "compact"})
+
+    def test_get_sample_size(self):
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:503]: new sample size check for tests

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 8]
```diff
@@ -493,12 +493,64 @@ def test_null_count(self):
...
+
+        # test data size * 0.20 < min_sample_size < data size
+        profile = dp.Profiler(data[:10000], profiler_options=profiler_options)
+        self.assertIn('sample_size', mocks[0].call_args[1])
+        self.assertEqual(5000, mocks[0].call_args[1]['sample_size'])
+
+        # test min_sample_size > data size * 0.20
+        profile = dp.Profiler(data, profiler_options=profiler_options)
+        self.assertIn('sample_size', mocks[0].call_args[1])
+        self.assertEqual(10000, mocks[0].call_args[1]['sample_size'])
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:552]: this test validates it is actually passed as expected

File: dataprofiler/profilers/profile_builder.py [hunk 9]
```diff
@@ -489,6 +475,22 @@ def __add__(self, other):
...
+        :type data: Union[data_readers.base_data.BaseData, pandas.DataFrame]
+        :return: integer sampling size
+        :rtype: int
+        """
+        len_data = len(data)
+        if self._samples_per_update:
+            return self._samples_per_update
+        if len_data <= self._min_sample_size:
+            return int(len_data)
+        return max(int(self._sampling_ratio * len_data), self._min_sample_size)
```
Comment (AnhTruong) [inline dataprofiler/profilers/profile_builder.py:492]: did we have a similar function for `StructuredDataProfile`? Can we use only one?

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 10]
```diff
@@ -493,12 +493,64 @@ def test_null_count(self):
...
+
+        # test data size * 0.20 < min_sample_size < data size
+        profile = dp.Profiler(pd.DataFrame([]))
+        sample_size = profile._get_sample_size(data[:10000])
+        self.assertEqual(5000, sample_size)
+
+        # test min_sample_size > data size * 0.20
+        profile = dp.Profiler(pd.DataFrame([]))
+        sample_size = profile._get_sample_size(data)
+        self.assertEqual(10000, sample_size)
```
Comment (AnhTruong) [inline dataprofiler/tests/profilers/test_profile_builder.py:519]: imo, for these tests we should set the value of self._min_sample_size and self._sampling_ratio and then test. Otherwise, if we change the default values later, the test will fail.

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 11]
```diff
@@ -570,6 +625,7 @@ def test_correct_total_sample_size_and_counts_and_mutability(self):
         file_path = os.path.join(test_root_path, 'data', 'csv/empty_rows.txt')
         data = pd.read_csv(file_path)
         profiler_options = ProfilerOptions()
+        profiler_options.structured_options.multiprocess.is_enabled = False
```
Comment (AnhTruong) [inline dataprofiler/tests/profilers/test_profile_builder.py:628]: so, we never test if the multiprocessing running for any test?

**Review by AnhTruong — COMMENTED** [review]
several comments

File: dataprofiler/profilers/profile_builder.py [hunk 12]
```diff
@@ -421,27 +418,16 @@ def __init__(self, data, samples_per_update=None, min_true_samples=0,
         self._min_true_samples = min_true_samples
         self._profile = dict()
 
+        # matches structured data profile
+        # TODO: allow set via options
+        self._sampling_ratio = 0.2
+        self._min_sample_size = 5000
```
Comment (lettergram) [inline dataprofiler/profilers/profile_builder.py:424]: Isn't this set by default already when you create the class?

File: dataprofiler/profilers/profile_builder.py (hunk 2 above)
Comment (lettergram) [reply dataprofiler/profilers/profile_builder.py:617]: It's not clear to me why we need to call `self._get_sample_size` within the `_update_profile_from_chunk`, we already pass in `sample_size` here.

File: dataprofiler/profilers/profile_builder.py [hunk 13]
```diff
@@ -642,14 +639,29 @@ def _update_profile_from_chunk(df, profile=None, sample_size=None,
...
+                data_labeler = DataLabeler(
+                    labeler_type='structured',
+                    dirpath=data_labeler_options.data_labeler_dirpath,
+                    load_options=None)
+                self.options.set(
+                    {'data_labeler.data_labeler_object': data_labeler})
+
+            except Exception as e:
+                utils.warn_on_profile('data_labeler', e)
+                self.options.set({'data_labeler.is_enabled': False})
```
Comment (lettergram) [inline dataprofiler/profilers/profile_builder.py:663]: I don't think this belongs in here, we don't instantiate anything in this function for good reason. It's `preprocessing` -> `processing` -> `postprocessing`, BUT the `processing` component isn't managed in here. IMO this convolutes the already hard to follow function

File: dataprofiler/profilers/profile_builder.py [hunk 14]
```diff
@@ -678,11 +693,11 @@ def tqdm(l):
         # Create structured profile objects
         new_cols = set()
         for col in df.columns:
-            if col not in profile:
+            if col not in self._profile:
```
Comment (lettergram) [inline dataprofiler/profilers/profile_builder.py:696]: I don't think any of these `self` are necessary. Also, I don't think you'd need to return profile if you were just doing it this way.

File: dataprofiler/tests/profilers/test_profile_builder.py (hunk 11 above)
Comment (lettergram) [reply dataprofiler/tests/profilers/test_profile_builder.py:628]: ^ I'm personally of the opinion multiprocessing should run on all tests, unless otherwise specified as it's the default behavior.

File: dataprofiler/profilers/profile_builder.py [hunk 15]
```diff
@@ -658,6 +670,9 @@ def tqdm(l):
                     print("Processing Column {}/{}".format(i+1, len(l)))
                     yield e
 
+        if not sample_size:
+            sample_size = self._get_sample_size(df)
+
```
Comment (lettergram) [inline dataprofiler/profilers/profile_builder.py:675]: I think this should be called outside the function and passed in

File: dataprofiler/profilers/profile_builder.py (hunk 9 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:492]: currently we use it in both spots, however this function is slightly different as it also utilizes the samples_per_update which is unique to this class. I did not want to alter the get_sample_size in `StructuredDataProfile` as it would prolong this fix

File: dataprofiler/profilers/profile_builder.py [hunk 16]
```diff
@@ -602,37 +604,32 @@ def update_profile(self, data, sample_size=None, min_true_samples=None):
         :type min_true_samples
         :return: None
         """
-        if not sample_size:
-            sample_size = self._samples_per_update
```
Comment (lettergram) [inline dataprofiler/profilers/profile_builder.py:606]: I think we should first check the data type, then get samples

File: dataprofiler/tests/profilers/test_profile_builder.py (hunk 10 above)
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_profile_builder.py:519]: fair

File: dataprofiler/tests/profilers/test_profile_builder.py (hunk 11 above)
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_profile_builder.py:628]: it requires significantly more time to run the tests. If we want to revery we can.

File: dataprofiler/profilers/profile_builder.py (hunk 12 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:424]: No, these are new to this class.

File: dataprofiler/profilers/profile_builder.py (hunk 13 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:663]: Technically this allows a profile to be instantiated without creating the labeler until profile time making it faster. Not sure if this is desirable.

However, where it was before puts the `processing` just before preprocessing anyways. In an ideal state, the data_labeler wouldn't be here anywhere in profiler anyways.

File: dataprofiler/profilers/profile_builder.py (hunk 14 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:696]: Could you explain further? this code no longer returns self._profile

File: dataprofiler/profilers/profile_builder.py [hunk 17]
```diff
@@ -768,9 +784,8 @@ def tqdm(l):
         print(notification_str)
         
         for col in tqdm(df.columns):
-            profile[col].update_column_profilers(clean_sampled_dict[col], pool)
```
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:771]: it is already changed?

File: dataprofiler/profilers/profile_builder.py (hunk 15 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:675]: moved

File: dataprofiler/profilers/profile_builder.py (hunk 16 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:606]: updated

**Review by gme5078 — APPROVED** [review]

**Review by lettergram — APPROVED** [review]
